File: src/main/scala/com/scala/study/akka/instruction/java/HelloWorldBot.java
package com.scala.study.akka.instruction.java;

import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;

public class HelloWorldBot extends AbstractBehavior<HelloWorld.Greeted> {
    private final int max;
    private int greetingCounter;

    public HelloWorldBot(ActorContext<HelloWorld.Greeted> context, int max) {
        super(context);
        this.max = max;
    }

    public static Behavior<HelloWorld.Greeted> create(int max) {
        return Behaviors.setup(context -> new HelloWorldBot(context, max));
    }


    @Override
    public Receive<HelloWorld.Greeted> createReceive() {
        return newReceiveBuilder()
                .onMessage(
                        HelloWorld.Greeted.class,
                        this::onGreeted
                )
                .build();
    }

    private Behavior<HelloWorld.Greeted> onGreeted(HelloWorld.Greeted message) {
        ++greetingCounter;
        getContext().getLog().info("Greeting {} for {} ", greetingCounter, message.whom);
        if (greetingCounter == max) {
            return Behaviors.stopped();
        } else {
            message.from.tell(new HelloWorld.Greet(message.whom, getContext().getSelf()));
            return this;
        }
    }
}

File: src/main/scala/com/scala/study/akka/instruction/scala/HelloWorldBot.scala
package com.scala.study.akka.instruction.scala

import akka.actor.typed.scaladsl.Behaviors
import akka.actor.typed.{ActorRef, ActorSystem, Behavior}

object HelloWorld{
  final case class Greet(whom:String,replyTo : ActorRef[Greeted])
  final case class Greeted(whom:String,from : ActorRef[Greet])

  def apply() : Behavior[Greet] = Behaviors.receive {
    (context,message) =>
      context.log.info("hello {} !",message.whom)
      message.replyTo ! Greeted(message.whom,context.self)
      Behaviors.same
  }
}


object HelloWorldBot {
  def apply(max : Int) : Behavior[HelloWorld.Greeted] = {
    bot(0,max)
  }

  private def bot(greetingCounter: Int,max : Int) : Behavior[HelloWorld.Greeted] = {
    Behaviors.receive{
      (context,message) =>
        val n = greetingCounter + 1;
        context.log.info("Greeting {} for {}",n,message.whom)
        if(n == max){
          Behaviors.stopped
        }else{
          message.from ! HelloWorld.Greet(message.whom,context.self)
          bot(n,max)
        }
    }
  }
}

object HelloWorldMain{
  final case class SayHello(name:String)
  def apply():Behavior[SayHello] = {
    Behaviors.setup{
      context =>
        val greeter = context.spawn(HelloWorld(),"greeter")
        Behaviors.receiveMessage{
          message =>
            context.log.info("HelloWorldMain message is  {}",message)
            val replyTo = context.spawn(HelloWorldBot(max = 3),message.name)
            greeter ! HelloWorld.Greet(message.name,replyTo)
            Behaviors.same
        }
    }
  }
}


object ActorSystemMain {
  def main(args: Array[String]): Unit = {
    val system : ActorSystem[HelloWorldMain.SayHello] = ActorSystem(HelloWorldMain(),"hello")
    system ! HelloWorldMain.SayHello("world")
  }
}